### PR TITLE
first attempt

### PR DESCRIPTION
**Pull request by terminalobject**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,32 +1,60 @@
-# Money tracker tech test
+Airport Challenge
```
Comment (maryrosecook) [inline README.md]: Not quite sure why this is here!

File: coverage/.last_run.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (maryrosecook) [inline coverage/.last_run.json:1]: Good job adding a test coverage checker!

Comment (maryrosecook) [inline coverage/.last_run.json:1]: Try to add each library or new set of files in a separate commit.  This makes it easier to roll back if you need to, and easier for others to read the commit history.

File: lib/account.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class Account
```
Comment (maryrosecook) [inline lib/account.rb:1]: Does this class have multiple responsibilities?

File: lib/account.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  end 
+
+  def add_entry(entry)
+    @entries << entry
+  end
+
+  def print_statement
+    puts 'date || credit || debit || balance'
+    @entries.each do |entry|
+      puts "#{entry.date} ||" + " ||" + " #{entry.debit} ||" + " #{balance_update(entry)}" if entry.credit.nil?
```
Comment (maryrosecook) [inline lib/account.rb]: `print_statement` is doing two things at once: generating the statement and updating the balance.  This makes it hard to reason about.

File: lib/account.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+      puts "#{entry.date} ||" + " #{entry.credit} ||" + " ||" + " #{balance_update(entry)}" if entry.debit.nil?
+    end
+  end
+  
+  private
+
+  def balance_update(entry)
```
Comment (maryrosecook) [inline lib/account.rb]: `balance_update` has multiple responsibilities.  Is there a way to refactor it to avoid this?

File: lib/account.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  def initialize(entries = [], balance = 2000)
+    @entries = entries
+    @balance = balance
+  end 
+
+  def add_entry(entry)
+    @entries << entry
+  end
+
+  def print_statement
```
Comment (maryrosecook) [inline lib/account.rb]: `print_statement` is a little hard to read because it is quite procedural.  Could it be made more declarative?

File: lib/account.rb (hunk 4 above)
Comment (maryrosecook) [inline lib/account.rb]: It's idiomatic in Ruby to avoid long lines.

File: lib/entry.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+require 'date'
+
+class Entry
+
+  attr_reader :date, :credit, :debit
+
+  def initialize(date, credit, debit)
+    raise ArgumentError.new 'At least one between credit and debit amount must not be nil.' if credit.nil? && debit.nil?
```
Comment (maryrosecook) [inline lib/entry.rb]: This line makes good use of ArgumentError!

File: lib/entry.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+require 'date'
+
+class Entry
+
+  attr_reader :date, :credit, :debit
+
+  def initialize(date, credit, debit)
```
Comment (maryrosecook) [inline lib/entry.rb:7]: Does this method have multiple responsibilities?

File: spec/account_spec.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
+require 'account'
+
+describe Account do
+  
+  let(:account) { Account.new([], 2000) }
+  let(:entry) { double(:entry) }
```
Comment (maryrosecook) [inline spec/account_spec.rb]: The `Account` tests make good use of doubles!

File: spec/account_spec.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  let(:second_entry) { double(:entry) }
```
Comment (maryrosecook) [inline spec/account_spec.rb]: It's really important to format code idiomatically.  Try using a linter to help with this.

File: spec/account_spec.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  it 'is initialized with a balance of 2000' do
+    expect(account.balance).to eq 2000
+  end
+
+  describe '#add_entry' do
+    it 'adds an entry to the entries array' do
+      account.add_entry(entry)
+      expect(account.entries.last).to eq entry
```
Comment (maryrosecook) [inline spec/account_spec.rb]: This is testing state, rather than behaviour.

File: spec/account_spec.rb [hunk 12]
```diff
@@ -0,0 +1,50 @@
+require 'account'
+
+describe Account do
+  
+  let(:account) { Account.new([], 2000) }
+  let(:entry) { double(:entry) }
+  let(:second_entry) { double(:entry) }
+  before(:each) do
```
Comment (maryrosecook) [inline spec/account_spec.rb]: This `before` block sets up one instance that's used by all the tests.  It's great to reuse setup code.  However, try to make it so the instance that a test uses has the minimal required setup done on it.

File: spec/account_spec.rb [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+  describe '#balance_update' do
+    it 'updates the balance according to the given entry amount' do
+      account.add_entry(entry)
+      account.balance_update(entry)
+      expect(account.balance).to eq 2500
+    end
+  end
+
+  describe '#print_statement' do
```
Comment (maryrosecook) [inline spec/account_spec.rb]: Try to test all the edge cases for each feature.

**Review by maryrosecook — COMMENTED** [review]
Here you go!

**Comment by maryrosecook** [comment]
Also, try to do lots of small commits as you go, rather than a few big commits.

File: README.md [hunk 14]
```diff
@@ -1,32 +1,12 @@
-# Money tracker tech test
+Money Tracker
```
Comment (maryrosecook) [inline README.md:1]: Good job writing user stories!

File: lib/entry.rb [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    @date, @credit, @debit = Date.parse(date), credit, debit unless !valid?(date, credit, debit)
+  end
+
+  def valid?(date, credit, debit)
```
Comment (maryrosecook) [inline lib/entry.rb:11]: Could this method be broken down to use smaller helper methods?

File: lib/entry.rb [hunk 16]
```diff
@@ -0,0 +1,27 @@
+require 'date'
+
+class Entry
+
+  attr_reader :date, :credit, :debit
+
+  def initialize(date, credit, debit)
+    @date, @credit, @debit = Date.parse(date), credit, debit unless !valid?(date, credit, debit)
```
Comment (maryrosecook) [inline lib/entry.rb:8]: This is a little hard to read.  Is the `unless` required, given that `valid?` throws if something is invalid? Could the checking the validity of the arguments be separated from the assignment of the arguments?

File: lib/statement.rb [hunk 17]
```diff
@@ -0,0 +1,38 @@
+class Statement 
+
+  attr_reader :tracker
+
+  def initialize(tracker = Tracker.new)
+    @tracker = tracker
+  end
+
+  def compose_line(date, credit, debit, balance)
```
Comment (maryrosecook) [inline lib/statement.rb:9]: Really nice method that's focused on one responsibility! One handy trick to avoid the `line[1]` stuff:

```
[date,
 credit.nil? ? "" : credit,
 debit.nil? ? "" : debit,
 balance].join(" ||  ")
```

File: lib/statement.rb [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    line = []
+    line[0] = date
+    line[1] = credit.nil? ? "" : credit
+    line[2] = debit.nil? ? "" : debit
+    line[3] = balance
+    line.join(" || ")
+  end
+
+  def compose_body
+    body = []
```
Comment (maryrosecook) [inline lib/statement.rb:19]: Is it possible to use `map` here to avoid having to explicitly make the body array?

File: lib/statement.rb [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    @tracker.account.entries.each do |entry|
+      body << compose_line(entry.date, entry.credit, entry.debit, @tracker.get_current_balance(@tracker.account, entry))
+    end 
+    body.join("\n")
+  end
+
+  def print_header
```
Comment (maryrosecook) [inline lib/statement.rb:26]: Could the `puts` in `print_header` and `print_body` be DRYed up?

File: lib/tracker.rb [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+  
+  attr_reader :initial_balance, :account
+
+  def initialize(account = Account.new([]), initial_balance = 0)
+    @account = account
+    @initial_balance = initial_balance
+  end 
+
+  def calculate_balance(account, entry)
+    current_balance = @initial_balance
```
Comment (maryrosecook) [inline lib/tracker.rb:11]: Really nice way of calculating the balance independent of iterating through the transactions!

File: lib/tracker.rb [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+    index = account.get_entry_index(entry)
+    (index + 1).times{ |k| current_balance += account.entries[k].net_amount }  
+    current_balance
+  end
+
+  def get_current_balance(account, entry)
+    return calculate_balance(account, entry)
```
Comment (maryrosecook) [inline lib/tracker.rb:18]: Is this `return` keyword needed?

File: lib/tracker.rb [hunk 22]
```diff
@@ -0,0 +1,20 @@
+class Tracker 
+  
+  attr_reader :initial_balance, :account
+
+  def initialize(account = Account.new([]), initial_balance = 0)
+    @account = account
+    @initial_balance = initial_balance
+  end 
+
+  def calculate_balance(account, entry)
```
Comment (maryrosecook) [inline lib/tracker.rb:10]: Does the balance calculation code belong on the `Tracker`? Or in another class?

File: spec/entry_spec.rb [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+  let(:entry) { Entry.new('2017-10-10', 500, nil) }
+  describe Entry.new('2017-03-03', 100, nil) do
+    it { is_expected.to respond_to(:date) } 
+    it { is_expected.to respond_to(:credit) }
+    it { is_expected.to respond_to(:debit) }
+  end
+
+  it 'is initialized with a date' do
+    expect(entry.date.to_s).to eq '2017-10-10'
+  end
```
Comment (maryrosecook) [inline spec/entry_spec.rb:14]: It's idiomatic to put whitespace between `its and `describes` and `lets`

File: spec/statement_spec.rb [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+describe Statement do
+
+  let(:date)      { '2017-02-02' }
+  let(:credit)    { 500 }
+  let(:debit)     { nil }
+  let(:balance)   { 1000 }
+  let(:entry)     { double :entry, date: '2017-06-06', credit: 1000, debit: nil}
+  let(:entry2)    { double :entry, date: '2017-06-06', credit: nil, debit: 500}
+  let(:tracker)   { double :tracker, account: account, initial_balance: 0 }
+  let(:account)   { double :account, entries: [entry, entry2] }
```
Comment (maryrosecook) [inline spec/statement_spec.rb:14]: Nice doubling!

**Review by maryrosecook — COMMENTED** [review]
Here you go!
